### PR TITLE
Design System: Menu for DropDown

### PR DESCRIPTION
**Pull request by BrittanyIRL**
## Summary

Adds menu functionality to `<DropDown />`. 

There's a variety of dropdown versions used throughout the dashboard and the editor currently. The idea here is to be able to decrease the amount of "one off" dropdowns in play by creating something with functionality that doesn't change but appearance that can.

In the new designs we have 2 types of dropdowns: 1) that has a single list and 2) one with sublists that will need to have sub list titles displayed. In each of these there can only be one option selected at a time.

Separately we have our typeahead/search dropdown, because the search functionality adds a good amount of complexity (lists that we should virtualize, the ability to have multiple values selected) this "version" of a dropdown shouldn't be considered a version at all, but it's own thing.

Lastly, we have popover menus which are only a thing in the dashboard right now, these are essentially just the "menu" part of the dropdown, these are separate from the dropdown too. As this is still a draft, I'm weighing pros vs cons of pulling out the useDropdownMenu hook I created to be shared.

## Relevant Technical Choices

-  `<DropDownMenu />` is responsible for the menu itself. It renders a list with its items by default. It can be overridden by including a renderItem component in <Dropdown />. There's examples of this in storybook and in tests. You can also override the styles by using menuStylesOverride with styled component's CSS prop. These overrides are here because it seems in our existing ecosystem one of the leading causes of making new dropDowns is the styling will vary (IE animation effects).

- `dropDown/menu/useDropDownMenu` takes over all of the keyboard functionality. In Menu we pass the hook props and it returns to us the focusedIndex (focusedValue is also available, but I've found it unnecessary to grab yet). Inside the hook we're finding the proper index to focus in the menu and focusing it and using the listRef that belongs to the <ul> inside of Menu to handle keyboard navigation. Having this logic separate from the UI should make things easier to share and avoid needless duplication of functionality.

## To-do

- [ ] Disabled item handling 
- [ ] Styles

## User-facing changes

None, changes live in storybook

## Testing Instructions

Run storybook locally and navigate to designsystem/components/dropdown. These dropdown demos should mimic our existing dropdown functionality.

- On default, if no value is currently selected the first list item should be focused
- On default, when a new item is selected the menu should close
- if isKeepMenuOpenOnSelection is true, the menu should not close when a new item is selected
- A menu should not open if the disabled knob is true in storybook
- 'No options available' should show up when menu is expanded if there are no items in an array
- If there is a currently selected value when a menu is opened it should be in focus and in view.
- You should be able to navigate dropdown list with your arrow keys. If isRTL knob is false then Up and Left arrows should move you up in the list. If isRTL knob is true then Up and Right arrows should move you up the list. (And respectfully, Down and Right arrows move you down, or Down and Left arrows move you down if isRTL is true).
- You should be able to select a new item by pressing Enter
- You should be able to close the menu by pressing Tab or Escape

---

Addresses #4950

### CODE REVIEW

File: assets/src/design-system/components/dropDown/index.js [hunk 1]
```diff
@@ -69,6 +86,31 @@ export const DropDown = ({
...
+  );
+
+  useEffect(() => {
+    if (selectRef?.current) {
+      const { height = null } = selectRef?.current?.getBoundingClientRect();
+      setAnchorHeight(height);
+    }
+  }, []);
+
+  const listId = `list-${uuidv4()}`;
```
Comment (maxyinger) [inline assets/src/design-system/components/dropDown/index.js]: does this need to be memoized at all so it's not producing a new id every render?

**Review by maxyinger — APPROVED** [review]
looks and works great! Excited to start integrating this more  into the codebase

File: assets/src/design-system/components/dropDown/menu/defaultListItem.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { Checkmark } from '../../../icons';
+import { DROP_DOWN_ITEM } from '../types';
+import { ListItem } from './components';
+
+export const DefaultListItem = forwardRef(function DefaultListItem(
```
Comment (carloskelly13) [inline assets/src/design-system/components/dropDown/menu/defaultListItem.js]: I could envision a `const` and `function` with the name name causing an issue. You could just `export default forwardRef(DefaultListItem)` then define `function DefaultListItem({ ... })` above that way we can assign `propTypes` to it.

File: assets/src/design-system/components/dropDown/menu/index.js [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+          {...groupAria}
+        >
+          {label && renderMenuLabel(label)}
+          {group.map((groupOption, optionIndex) =>
+            renderMenuItem(groupOption, optionIndex, groupIndex)
+          )}
+        </List>
+      );
+    });
+  }, [options, emptyText, renderMenuItem, listId]);
```
Comment (carloskelly13) [inline assets/src/design-system/components/dropDown/menu/index.js]: In this functional component we have `MenuContent`, `renderMenuItem`, `renderMenuLabel`. I think they are big and complex enough to become their own components. The latter two could turn into components like `<MenuItem />` and `<MenuLabel />` with props too. While they are memoized, they still need to be checked on each render. This should also make them more composable.

**Review by carloskelly13 — CHANGES_REQUESTED** [review]
This is great and really covers a bunch of the use cases for how we use drop down menus. I do think because of the complexity of our use cases we should further chop up the `DropDownMenu` component into smaller components.

File: assets/src/design-system/components/dropDown/menu/defaultListItem.js (hunk 2 above)
Comment (BrittanyIRL) [reply assets/src/design-system/components/dropDown/menu/defaultListItem.js]: Sounds good!

File: assets/src/design-system/components/dropDown/menu/index.js (hunk 3 above)
Comment (BrittanyIRL) [reply assets/src/design-system/components/dropDown/menu/index.js]: Thanks for calling this out, I was waffling between this. Updating now :)

Comment (maxyinger) [reply assets/src/design-system/components/dropDown/menu/index.js]: Haah I was on the fence about updating this too. glad it got called out

Comment (BrittanyIRL) [reply assets/src/design-system/components/dropDown/menu/index.js]: @carlos-kelly @littlemilkstudio thoughts about this update? 

I pulled apart anything that was rendering into it's own component and file. So `/index` is now more of a container that gives functionality to the list. If there's no options them the `EmptyList` component takes over. Otherwise, we have `ListGroupings` which then sets up a `List` with a `GroupLabel` if relevant per group and then proceeds to map through options, where if `renderItem` was passed into the `<DropDown />` it will override `ListItem` but otherwise will default to use `menu/defaultListItem`. 

I think this gives us more room to test edge cases in the menu itself and I think the naming makes it clear what's happening.

Comment (maxyinger) [reply assets/src/design-system/components/dropDown/menu/index.js]: I love it. Mush easier to read and reason about. Great work!

File: assets/src/design-system/components/dropDown/menu/listGroupings.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  activeValue,
+  listLength,
+  listId,
+  hasMenuRole,
+  handleMenuItemSelect,
+  renderItem,
+  optionsRef,
+}) => {
+  const ListItem = renderItem || DefaultListItem;
+  const isManyGroups = useMemo(() => options?.length > 1, [options.length]);
```
Comment (carloskelly13) [inline assets/src/design-system/components/dropDown/menu/listGroupings.js:42]: Since this is just testing if something is > than a value, it doesn't need to be memoized. The comparison could be more expensive than the test. Alternatively if this only changes once within the lifecycle of the ListGrouping, we could put this in a `useRef` like `const isManyGroups = useRef(Boolean(options?.length > 1)).current`

**Review by carloskelly13 — APPROVED** [review]
